Plugin mono repo

6utt3rfly opened the pull request:
- Adds combined plugin tests
- Moves to mono-repo type structure for plugins. Each plugin has to be separately published to npm (all are starting at 1.0.0). I've named them `@jsep/plugin-object` and so-on.
- (the change is a lot of new files copied and pasted as boiler-plate code for each plugin, but I moved the existing plugins and tests without changing the code)

```
├── README.md
├── package.json
├── plugins
│   ├── plugin-assignment
│   │   ├── dist
│   │   │   ├── cjs
│   │   │   │   ├── index.cjs.js
│   │   │   └── index.js
│   │   ├── src
│   │   │   └── index.js
│   │   ├── test
│   │   │   └── index.test.js
│   │   ├── types
│   │   │   └── ts.d.ts
│   │   ├── package.json
│   │   ├── README.md
│   │   └── rollup.config.js
│   └── ...plugins
├── src
│   ├── jsep.js
│   └── index.js
├── test
│   ├── jsep.test.js
├── rollup.config.js
```

I added a `build:all` script, that is just `cd plugins/{name} && npm run build...`
I manually copied everything to a typescript project and was able to:
```typescript
import * as jsep from 'jsep';
import * as jsepArrow from '@jsep/plugin-arrow';
jsep.plugins.register(jsepArrow);
```

_Note: I've never created a mono-repo before, but I copied rollup as an example framework and I think it works..._

6utt3rfly committed:
Add combined plugins tests to ensure cross-compatibility when using multiple plugins

6utt3rfly committed:
Move to a mono-repo type structure for plugins. Each plugin is its own folder with its own package.json, rollup, etc. This keeps them more independent of jsep and allows for separate versioning

LeaVerou commented:
I'm not very familiar with how mono-repos work. Maybe @EricSmekens should review this one.

I do like that plugins seem to be entirely self contained.
I don't like the repetition. It looks like if we make a change to the build process, we'd need to make it N times, where N is the number of plugins? Is there a way to make this a little more DRY?

How does versioning work in a monorepo? Do plugins have their own versions or does the entire repo have a single version?

6utt3rfly committed:
Reduce boilerplate:
- remove .gitignore (root file works)
- remove rollup.config.js (and use a single file at the root)

Add plugin-specific types

Fix New plugin adding 'new' to a MemberExpression instead of the CallExpression

6utt3rfly commented:
I removed a couple files per repo, but there are still 7 files:
- src file
- test file
- readme
- package.json (could potentially be generated, unless a plugin needed a special package)
- types (sometimes common, but I've added plugin-specific extensions of `Expression` to most of them)
- unit_tests.html -- could be generated (or I think qunit can be run without a browser)
- .npmignore could be removed if we wrote a script to publish to npm or don't care about src and test folders?

Each plugin would have to be separately published to npm (that's why each has it's own package.json), and so they have independent version numbers. I don't know if there's anything special about publishing to `@{string}/...` (in this case `@jsep/plugin-{name}` or if it's first-come-first-serve like usual?

EricSmekens commented:
I like the well documented, tested and extensible changes that were made.
I don't have any experience with mono-repo/multiple package repo's like this either, but I think we'll manage this with deploying and learn from the initial publishes.

There are some downsides to it, but being able to patch/extend 1 plugin is a advantage. I cannot weigh them properly at this moment, but I feel like this is a fine way ahead and we can make changes along the way.

(Feel free to merge this into your existing PR, and mention me again when we want to move this to npm!)

(Also, apologies for the time I take to reply, busy times).

6utt3rfly committed:
Rename `plugins` folder to `packages` and update references
